File: MetaGeta/MetaGeta.GUI/ViewModels/DataStoreConfigurationViewModel.vb
Public Class DataStoreConfigurationViewModel
    Inherits NavigationTab

    Private ReadOnly m_DataStore As MGDataStore
    Private m_SelectedPlugin As IMGPluginBase

    Public Sub New(ByVal dataStore As MGDataStore)
        m_DataStore = dataStore
        AddHandler m_DataStore.PropertyChanged, AddressOf DataStore_PropertyChanged

        m_DeleteCommand = New RelayCommand(AddressOf Delete)
    End Sub


    Private m_DeleteCommand As RelayCommand
    Public ReadOnly Property DeleteCommand() As ICommand
        Get
            Return m_DeleteCommand
        End Get
    End Property
    Private Sub Delete()
        m_DataStore.Delete()
    End Sub

    Public Property Name() As String
        Get
            Return m_DataStore.Name
        End Get
        Set(ByVal value As String)
            m_DataStore.Name = value
        End Set
    End Property

    Public Property Description() As String
        Get
            Return m_DataStore.Description
        End Get
        Set(ByVal value As String)
            m_DataStore.Description = value
        End Set
    End Property

    Public ReadOnly Property Plugins() As IEnumerable(Of IMGPluginBase)
        Get
            Return m_DataStore.Plugins
        End Get
    End Property
    Public Property SelectedPlugin() As IMGPluginBase
        Get
            Return m_SelectedPlugin
        End Get
        Set(ByVal value As IMGPluginBase)
            If m_SelectedPlugin IsNot value Then
                m_SelectedPlugin = value
                OnPropertyChanged("SelectedPlugin")
                OnPropertyChanged("SelectedPluginSettings")
            End If
        End Set
    End Property

    Private Sub DataStore_PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
        If e.PropertyName = "Name" Then
            OnPropertyChanged("Name")
        ElseIf e.PropertyName = "Description" Then
            OnPropertyChanged("Description")
        End If
    End Sub

    Public ReadOnly Property SelectedPluginSettings() As SettingInfoCollection
        Get
            If SelectedPlugin Is Nothing Then
                Return Nothing
            Else
                Return SettingInfoCollection.GetSettings(CType(SelectedPlugin, IMGPluginBase))
            End If
        End Get
    End Property

    Public Overrides ReadOnly Property Caption() As String
        Get
            Return "Configuration"
        End Get
    End Property

    Public Overrides ReadOnly Property Icon() As System.Windows.Media.ImageSource
        Get
            Return s_RunImage
        End Get
    End Property

    Private Shared ReadOnly s_RunImage As New BitmapImage(New Uri("pack://application:,,,/MetaGeta.GUI;component/Resources/run.png"))
End Class

File: MetaGeta/MetaGeta.GUI/ViewModels/DataStoreConfigurationViewModel.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using MetaGeta.DataStore;

#endregion

namespace MetaGeta.GUI {
    public class DataStoreConfigurationViewModel : NavigationTab {
        private static readonly BitmapImage s_RunImage = new BitmapImage(new Uri("pack://application:,,,/MetaGeta.GUI;component/Resources/run.png"));
        private readonly MGDataStore m_DataStore;
        private readonly RelayCommand m_DeleteCommand;

        private IMGPluginBase m_SelectedPlugin;

        public DataStoreConfigurationViewModel(MGDataStore dataStore) {
            m_DataStore = dataStore;
            m_DataStore.PropertyChanged += DataStore_PropertyChanged;

            m_DeleteCommand = new RelayCommand(Delete);
        }


        public ICommand DeleteCommand {
            get { return m_DeleteCommand; }
        }

        public string Name {
            get { return m_DataStore.Name; }
            set { m_DataStore.Name = value; }
        }

        public string Description {
            get { return m_DataStore.Description; }
            set { m_DataStore.Description = value; }
        }

        public IEnumerable<IMGPluginBase> Plugins {
            get { return m_DataStore.Plugins; }
        }

        public IMGPluginBase SelectedPlugin {
            get { return m_SelectedPlugin; }
            set {
                if (!ReferenceEquals(m_SelectedPlugin, value)) {
                    m_SelectedPlugin = value;
                    OnPropertyChanged("SelectedPlugin");
                    OnPropertyChanged("SelectedPluginSettings");
                }
            }
        }

        public SettingInfoCollection SelectedPluginSettings {
            get {
                if (SelectedPlugin == null)
                    return null;
                else
                    return SettingInfoCollection.GetSettings(SelectedPlugin);
            }
        }

        public override string Caption {
            get { return "Configuration"; }
        }

        public override ImageSource Icon {
            get { return s_RunImage; }
        }

        private void Delete() {
            m_DataStore.Delete();
        }

        private void DataStore_PropertyChanged(object sender, PropertyChangedEventArgs e) {
            if (e.PropertyName == "Name")
                OnPropertyChanged("Name");
            else if (e.PropertyName == "Description")
                OnPropertyChanged("Description");
        }
    }
}